GoogleSignIn/Sources/GIDGoogleUser.m

```diff
@@ -101,44 +106,128 @@ - (GIDConfiguration *)configuration {
+      });
+    }
+#endif // TARGET_OS_IOS && !TARGET_OS_MACCATALYST
+  }];
+}
+
+- (OIDAuthState *) authState{
+  return ((GTMAppAuthFetcherAuthorization *)self.fetcherAuthorizer).authState;
+}
+
 #pragma mark - Private Methods
 
 #if TARGET_OS_IOS && !TARGET_OS_MACCATALYST
 - (nullable NSString *)emmSupport {
-  return
-      _authState.lastAuthorizationResponse.request.additionalParameters[kEMMSupportParameterName];
+  return self.authState.lastAuthorizationResponse
+      .request.additionalParameters[kEMMSupportParameterName];
 }
 #endif // TARGET_OS_IOS && !TARGET_OS_MACCATALYST
 
 - (instancetype)initWithAuthState:(OIDAuthState *)authState
                       profileData:(nullable GIDProfileData *)profileData {
   self = [super init];
   if (self) {
-    [self updateAuthState:authState profileData:profileData];
-  }
-  return self;
-}
-
-- (void)updateAuthState:(OIDAuthState *)authState
-            profileData:(nullable GIDProfileData *)profileData {
-  @synchronized(self) {
-    _authState = authState;
-    _authentication = [[GIDAuthentication alloc] initWithAuthState:authState];
+    _tokenRefreshHandlerQueue = [[NSMutableArray alloc] init];
     _profile = profileData;
     
 #if TARGET_OS_IOS && !TARGET_OS_MACCATALYST
     GTMAppAuthFetcherAuthorization *authorization = self.emmSupport ?
-        [[GIDAppAuthFetcherAuthorizationWithEMMSupport alloc] initWithAuthState:_authState] :
-        [[GTMAppAuthFetcherAuthorization alloc] initWithAuthState:_authState];
+        [[GIDAppAuthFetcherAuthorizationWithEMMSupport alloc] initWithAuthState:authState] :
+        [[GTMAppAuthFetcherAuthorization alloc] initWithAuthState:authState];
 #elif TARGET_OS_OSX || TARGET_OS_MACCATALYST
     GTMAppAuthFetcherAuthorization *authorization =
-        [[GTMAppAuthFetcherAuthorization alloc] initWithAuthState:_authState];
+        [[GTMAppAuthFetcherAuthorization alloc] initWithAuthState:authState];
 #endif // TARGET_OS_IOS && !TARGET_OS_MACCATALYST
     authorization.tokenRefreshDelegate = self;
+    authorization.authState.stateChangeDelegate = self;
     self.fetcherAuthorizer = authorization;
     
     [self updateTokensWithAuthState:authState];
   }
+  return self;
+}
+
+- (void)updateWithTokenResponse:(OIDTokenResponse *)tokenResponse
+          authorizationResponse:(OIDAuthorizationResponse *)authorizationResponse
+                    profileData:(nullable GIDProfileData *)profileData {
+  @synchronized(self) {
+    _profile = profileData;
+    
+    // We don't want to trigger the delegate before we update authState completely. So we unset the
+    // delegate before the first update. Also the order of updates is important because
+    // `updateWithAuthorizationResponse` would clears the last token reponse and refresh token.
+    self.authState.stateChangeDelegate = nil;
+    [self.authState updateWithAuthorizationResponse:authorizationResponse error:nil];
+    self.authState.stateChangeDelegate = self;
```

Comment:
AuthState doesn't provide an API to update itself with both authorization response and token response. We have to do this in two steps and every time we update the delegate would be triggered. (We are considering adding this method in AppAuth in the future.)
If we want to trigger the delegate once this is the workaround. Otherwise we can also accept this and update our tokens twice.

Reply:
Looks like `updateTokensWithAuthState:` already does equality check before set new tokens, so I don't see any harm in calling it twice in this case. And unsetting the delegate temporarily feels a little too hacky. So I'd suggest against unsetting and resetting the delegate here.

Reply:
Created an issue in AppAuth. https://github.com/openid/AppAuth-iOS/issues/728

Reply:
Please link to the AppAuth issue in the comment above the code.

Reply:
Done.